compilation/src/main/scala/scala/tools/nsc/HotSbtBenchmark.scala

```diff
@@ -0,0 +1,141 @@
+package scala.tools.nsc
+
+import java.io._
+import java.net.URL
+import java.nio.file._
+import java.nio.file.attribute.BasicFileAttributes
+import java.util.concurrent.TimeUnit
+
+import com.typesafe.config.ConfigFactory
+import org.openjdk.jmh.annotations.Mode.SampleTime
+import org.openjdk.jmh.annotations._
+
+import scala.collection.JavaConverters._
+
+@State(Scope.Benchmark)
+@BenchmarkMode(Array(SampleTime))
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 10, time = 10, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 10, time = 10, timeUnit = TimeUnit.SECONDS)
+@Fork(value = 3)
+class HotSbtBenchmark {
+  @Param(value = Array())
+  var source: String = _
+
+  @Param(value = Array(""))
+  var extraArgs: String = _
+
+  // This parameter is set by ScalacBenchmarkRunner / UploadingRunner based on the Scala version.
+  // When running the benchmark directly the "latest" symlink is used.
+  @Param(value = Array("latest"))
+  var corpusVersion: String = _
+
+  var sbtProcess: Process = _
+  var inputRedirect: ProcessBuilder.Redirect = _
+  var outputRedirect: ProcessBuilder.Redirect = _
+  var tempDir: Path = _
+  var scalaHome: Path = _
+  var processOutputReader: BufferedReader = _
+  var processInputReader: BufferedWriter = _
+  var output= new java.lang.StringBuilder()
+
+  def buildDef =
+    s"""
+       |scalaHome := Some(file("${scalaHome.toAbsolutePath.toString}"))
+       |
+       |val cleanClasses = taskKey[Unit]("clean the classes directory")
+       |
+       |cleanClasses := IO.delete((classDirectory in Compile).value)
+       |
+       |scalaSource in Compile := file("${corpusSourcePath.toAbsolutePath.toString}")
+       |
+       |libraryDependencies += "org.scala-lang" % "scala-compiler" % scalaVersion.value
+       |libraryDependencies += "org.scala-lang" % "scala-reflect" % scalaVersion.value
+       |
+       |// TODO support .java sources
+    """.stripMargin
+
+  @Setup(Level.Trial) def spawn(): Unit = {
+    tempDir = Files.createTempDirectory("sbt-")
+    scalaHome = Files.createTempDirectory("scalaHome-")
+    initDepsClasspath()
+    Files.createDirectory(tempDir.resolve("project"))
+    Files.write(tempDir.resolve("project/build.properties"), java.util.Arrays.asList("sbt.version=0.13.15"))
+    Files.write(tempDir.resolve("build.sbt"), buildDef.getBytes("UTF-8"))
+    val sbtLaucherPath = System.getProperty("sbt.launcher")
+    if (sbtLaucherPath == null) sys.error("System property -Dsbt.launcher absent")
+    val builder = new ProcessBuilder(sys.props("java.home") + "/bin/java", "-Xms2G", "-Xmx2G", "-Dsbt.log.format=false", "-jar", sbtLaucherPath)
```

Comment:
- [ ] Disable JLine in the spawned SBT process to avoid hosing the ambient terminal sometimes.